pkg/tfbridge/detailed_diff.go

```diff
@@ -0,0 +1,406 @@
+}
+
+type baseDiff string
+
+const (
+	NoDiff    baseDiff = "NoDiff"
+	Add       baseDiff = "Add"
+	Delete    baseDiff = "Delete"
+	Update    baseDiff = "Update"
+	Undecided baseDiff = "Undecided"
+)
+
+func (b baseDiff) ToPropertyDiff() *pulumirpc.PropertyDiff {
+	contract.Assertf(b != Undecided, "diff should not be undecided")
+	switch b {
+	case Add:
+		return &pulumirpc.PropertyDiff{Kind: pulumirpc.PropertyDiff_ADD}
+	case Delete:
+		return &pulumirpc.PropertyDiff{Kind: pulumirpc.PropertyDiff_DELETE}
+	case Update:
+		return &pulumirpc.PropertyDiff{Kind: pulumirpc.PropertyDiff_UPDATE}
+	default:
+		return nil
+	}
+}
+
+func makeBaseDiff(old, new resource.PropertyValue) baseDiff {
+	oldPresent := isPresent(old)
+	newPresent := isPresent(new)
+	if !oldPresent {
+		if !newPresent {
+			return NoDiff
+		}
+
+		return Add
+	}
+	if !newPresent {
+		return Delete
+	}
+
+	if new.IsComputed() {
+		return Update
+	}
+
+	return Undecided
+}
+
+type (
+	detailedDiffKey string
+	propertyPath    resource.PropertyPath
+)
+
+func newPropertyPath(root resource.PropertyKey) propertyPath {
+	return propertyPath{string(root)}
+}
+
+func (k propertyPath) String() string {
+	return resource.PropertyPath(k).String()
+}
+
+func (k propertyPath) Key() detailedDiffKey {
+	return detailedDiffKey(k.String())
+}
+
+func (k propertyPath) append(subkey interface{}) propertyPath {
+	return append(k, subkey)
+}
+
+func (k propertyPath) Subpath(subkey string) propertyPath {
+	return k.append(subkey)
+}
+
+func (k propertyPath) Index(i int) propertyPath {
+	return k.append(i)
+}
+
+func (k propertyPath) IsReservedKey() bool {
+	leaf := k[len(k)-1]
+	return leaf == "__meta" || leaf == "__defaults"
+}
+
+func mapHasReplacements(m map[detailedDiffKey]*pulumirpc.PropertyDiff) bool {
+	for _, diff := range m {
+		if diff.GetKind() == pulumirpc.PropertyDiff_ADD_REPLACE ||
+			diff.GetKind() == pulumirpc.PropertyDiff_DELETE_REPLACE ||
+			diff.GetKind() == pulumirpc.PropertyDiff_UPDATE_REPLACE {
+			return true
+		}
+	}
+	return false
+}
+
+type detailedDiffer struct {
+	tfs shim.SchemaMap
+	ps  map[string]*SchemaInfo
+}
+
+func (differ detailedDiffer) propertyPathToSchemaPath(path propertyPath) walk.SchemaPath {
+	return PropertyPathToSchemaPath(resource.PropertyPath(path), differ.tfs, differ.ps)
+}
+
+func (differ detailedDiffer) getEffectiveType(path walk.SchemaPath) shim.ValueType {
+	tfs, ps, err := LookupSchemas(path, differ.tfs, differ.ps)
+
+	if err != nil || tfs == nil {
+		return shim.TypeInvalid
+	}
+
+	if IsMaxItemsOne(tfs, ps) {
+		return differ.getEffectiveType(path.Element())
+	}
+
+	return tfs.Type()
+}
+
+func (differ detailedDiffer) lookupSchemas(path propertyPath) (shim.Schema, *info.Schema, error) {
+	schemaPath := PropertyPathToSchemaPath(resource.PropertyPath(path), differ.tfs, differ.ps)
+	return LookupSchemas(schemaPath, differ.tfs, differ.ps)
+}
+
+func (differ detailedDiffer) isForceNew(pair propertyPath) bool {
+	// See pkg/cross-tests/diff_cross_test.go
```

Comment:
This is probably fine but looks a little complicated so might be good to explain in plain words in addition to linking to test cases. Why is it going up the parent chain?